tools/mafft.py

```diff
@@ -0,0 +1,157 @@
+'''
+    MAFFT - Multiple alignment program for amino acid or nucleotide sequences
+    http://mafft.cbrc.jp/alignment/software/
+'''
+
+__author__ = "tomkinsc@broadinstitute.org"
+
+import logging, tools, util.file
+import os, os.path, subprocess
+
+tool_version = '7.220'
+url = 'http://mafft.cbrc.jp/alignment/software/mafft-{ver}-{os}.{ext}'
+
+log = logging.getLogger(__name__)
+
+class MafftTool(tools.Tool):
+    def __init__(self, install_methods = None):
+        if install_methods == None:
+            install_methods = []
+            mafft_os                = get_mafft_os()
+            mafft_bitdepth          = get_mafft_bitdepth()
+            mafft_archive_extension = get_mafft_archive_extension(mafft_os)
+            binaryPath              = get_mafft_binary_path(mafft_os, mafft_bitdepth)
+            binaryDir               = get_mafft_binary_path(mafft_os, mafft_bitdepth, full=False)
+            
+            target_rel_path = '{binPath}'.format(ver=tool_version, os=mafft_os, binPath=binaryPath)
+            verify_command  = 'cd {dir}/mafft-{ver}/{binDir} && {dir}/mafft-{ver}/{binPath} --version > /dev/null 2>&1'.format(dir=util.file.get_build_path(), ver=tool_version, os=mafft_os, binPath=binaryPath, binDir=binaryDir) 
+            destination_dir = '{dir}/mafft-{ver}'.format(dir=util.file.get_build_path(), ver=tool_version, os=mafft_os, binPath=binaryPath)
+
+            install_methods.append(
+                    tools.DownloadPackage( url.format(ver=tool_version, os=mafft_os, ext=mafft_archive_extension),
+                    target_rel_path=target_rel_path,
+                    destination_dir=destination_dir,
+                    verifycmd=verify_command))
+
+        tools.Tool.__init__(self, install_methods = install_methods)
+
+    def version(self):
+        return tool_version
+
+    def execute(self, inFastas, outFile, localpair, globalpair, preservecase, reorder, 
+                outputAsClustal, maxiters, gapOpeningPenalty=None, offset=None, threads=1, verbose=True):
+
+        inputFileName         = ""
+        tempCombinedInputFile = ""
+
+        # get the full paths of input and output files in case the user has specified relative paths
+        inputFiles = []
+        for f in inFastas:
+            inputFiles.append(os.path.abspath(f))
+        outFile = os.path.abspath(outFile)
+
+        # if multiple fasta files are specified for input
+        if len(inputFiles)>1:
+            # combined specified input files into a single temp FASTA file so MAFFT can read them
+            tempFileSuffix = ""
+            for filePath in inputFiles:
```

Comment:
The first block gets the absolute filepath for each file specified since the subprocess pwd can vary by platform and relative paths can fail. The next bock concatenates multiple FASTA files into one temp file since MAFFT expects a single file as input.